neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+- The sign value is not 0 or 1
+
+***Annotation:***
+
+```rust
+/// Computes sum for signed elements on BLS12-381 curve 
+/// \sum_i (-1)^{s_i} p_i should be the equal result.
+///
+/// # Arguments
+///
+/// * `value` - sequence of (pi, si), where pi is point(x:Fp^2, y:Fp^2) on BLS-381 curve and si is sign (0 for +, 1 for -) 
+///    BLS12-381 is Y^2 = X^3 + 4(i + 1) curve over Fp^2.
+///
+///   `value` is encoded as packed, big-endian
+///   `[(([u8; 96], [u8; 96]), u8)]` slice.
+///
+/// # Errors
+///
+/// If `value_len + value_ptr` points outside the memory or the registers
+/// use more memory than the limit, the function returns `MemoryAccessViolation`.
+///
+/// If (1) point coordinates are not on curve or (2) `value.len()%193 != 0` or (3) sign is not 0 or 1,
+/// the function returns `BLS12381InvalidInput`.
+///
+/// # Cost
+///
+/// `base + write_register_base + write_register_byte * num_bytes +
+/// bls12381_g2_sum_base + bls12381_g2_sum_element * num_elements`
+pub fn bls12381_g2_sum(&mut self, 
+                       value_len: u64, 
+                       value_ptr: u64, 
+                       register_id: u64) -> Result<()>;
+```
+
+#### ***bls12381_g1_multiexp***
+
+***Description:***
+
+The function takes as input the list of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is a point on the curve and $s_i \in \mathbb{N}_0$ is a scalar. The function calculates $\sum s_i \cdot p_i$.
+
+The multiplication on the scalar is the addition of that point a scalar number of times:
+
+$$
+s \cdot p = \underbrace{p + p + \ldots + p}_{s}
+$$
+
+The $E(F_p)$ curve,  points on the curve and the addition operation are defined in the BLS12-381 Curve Specification section.
+
+Note:
+
+- We take as input any points on the curve, not only from   $G_1$.
+- The scalar is an arbitrary unsigned integer and can be bigger than the group order.
+
+***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is a point on the curve and $s_i \in \mathbb{N}_0$ is a scalar. Each point is encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and a scalar has a type `u256` and BigEndian encoded in 32 bytes. Expected `128*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the concatenation of uncompressed point from $E(F_p)$ — `96` bytes and scalar — `32` bytes. More details are in the Curve Points Encoding section.
+
+***Output:*** the output is 96 bytes — the one point $\in E(F_p)$ in decompressed form. For empty input it returns point on infinity (see  Curve Points Encoding section).
+
+***Gas Estimation:***
+
+This function should be calculated by Pippenger’s algorithm[^25]. The Complexity of this algorithm is $O(\frac{k}{\log(k)})$. For gas calculation we will use the formula $\frac{k}{\max(\log_2(k), 1)}$ the same way as in the host function for `BN254`[^10].
+
+```rust
+let k = (input_bytes+item_size-1)/item_size;
+let gas_consumed = A + B*k + C * if k > 1 {k / (k as f32).log2().floor()} else {k};
```

Comment:
@Ekleog-NEAR From the algorithmic point of view, it is a correct formula. I took this from bn254 NEP. However, after I started implementing this, and figured out, that for bn254 this formula is not implemented, instead just a linear formula is used. 

The idea of that formula: the algorithm gives some speed boost and for using this boost we need this formula. Otherwise, we will charge too much gas on the big number of points. 

In EIP-2537 they have just a list of constants of how much gas they should charge for different N. 

Now, I have started to think, that the best solution:
- using the linear formula for gas calculation
- don't specify which algorithm should be used  in the implementation

The Pippenger algorithm is implemented only in the master branch of blst library, not in the release version. And I started to doubt that implementation is safe. As a result, I think to make the first implementation as simple as possible and don't use a clever algorithm. In the future, if we will need it, we can update it to the faster version. We can update it without changing this NEP. And I expect that for this function, slow implementation will be acceptable.

Reply:
fixed: https://github.com/near/NEPs/pull/488/commits/80adcc7167832afd653386381d621b77a01d1f4f